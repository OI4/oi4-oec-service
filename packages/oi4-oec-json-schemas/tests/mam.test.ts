import { matchersWithOptions } from 'jest-json-schema'
import schema from '../schemas/mam.schema.json'
import LocalizedText from '../schemas/LocalizedText.schema.json'
import int32 from '../schemas/dataTypes/int32.schema.json'
import localeSchema from '../schemas/constants/locale.pattern.schema.json'

import validObjs from './__fixtures__/mam_valid.json'
import invalidObjs from './__fixtures__/mam_invalid.json'

expect.extend(
  matchersWithOptions({
    // Loading in a schema which is comprised only of definitions,
    // which means specific test schemas need to be created.
    // This is good for testing specific conditions for definition schemas.
    schemas: [LocalizedText, int32],
    verbose: true,
  }, (ajv) => {
    ajv.addSchema(localeSchema)
  })
)

describe('mam schema', () => {
  it('validate schema', () => {
    expect(schema).toBeValidSchema()
  })

  it.each(validObjs as [])(
    '(%#) match valid config object to schema -> %s',
    (_name: string, obj) => {
      expect(obj).toMatchSchema(schema)
    }
  )

  it.each(invalidObjs as [])(
    '(%#) match fails for invalid config -> %s',
    (name: string, obj) => {
      expect(obj).not.toMatchSchema(schema)
      expect(() => {
        expect(obj).toMatchSchema(schema)
      }).toThrowErrorMatchingSnapshot(name)
    }
  )
})
