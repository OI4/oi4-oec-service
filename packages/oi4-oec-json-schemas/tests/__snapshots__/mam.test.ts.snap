// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mam schema (0) match fails for invalid config -> empty: empty 1`] = `
"expect(received).toMatchSchema(schema)

received
  should have required property 'DeviceClass'
    Path:     mam.schema.json#/required
  should have required property 'DeviceManual'
    Path:     mam.schema.json#/required
  should have required property 'DeviceRevision'
    Path:     mam.schema.json#/required
  should have required property 'HardwareRevision'
    Path:     mam.schema.json#/required
  should have required property 'Manufacturer'
    Path:     mam.schema.json#/required
  should have required property 'ManufacturerUri'
    Path:     mam.schema.json#/required
  should have required property 'Model'
    Path:     mam.schema.json#/required
  should have required property 'ProductCode'
    Path:     mam.schema.json#/required
  should have required property 'ProductInstanceUri'
    Path:     mam.schema.json#/required
  should have required property 'RevisionCounter'
    Path:     mam.schema.json#/required
  should have required property 'SerialNumber'
    Path:     mam.schema.json#/required
  should have required property 'SoftwareRevision'
    Path:     mam.schema.json#/required
"
`;

exports[`mam schema (1) match fails for invalid config -> null values: null values 1`] = `
"expect(received).toMatchSchema(schema)

received
  .Description should be object
    Received: <null>
    Path:     mam.schema.json#/type
  .DeviceClass should be string
    Received: <null>
    Path:     mam.schema.json#/properties/DeviceClass/type
  .DeviceManual should be string
    Received: <null>
    Path:     mam.schema.json#/properties/DeviceManual/type
  .DeviceRevision should be string
    Received: <null>
    Path:     mam.schema.json#/properties/DeviceRevision/type
  .HardwareRevision should be string
    Received: <null>
    Path:     mam.schema.json#/properties/HardwareRevision/type
  .Manufacturer.locale should be string
    Received: <null>
    Path:     mam.schema.jsonlocale.pattern.schema.json/type
  .Manufacturer.text should be string
    Received: <null>
    Path:     mam.schema.json#/properties/text/type
  .ManufacturerUri should be string
    Received: <null>
    Path:     mam.schema.json#/properties/ManufacturerUri/type
  .Model.locale should be string
    Received: <null>
    Path:     mam.schema.jsonlocale.pattern.schema.json/type
  .Model.text should be string
    Received: <null>
    Path:     mam.schema.json#/properties/text/type
  .ProductCode should be string
    Received: <null>
    Path:     mam.schema.json#/properties/ProductCode/type
  .ProductInstanceUri should be string
    Received: <null>
    Path:     mam.schema.json#/properties/ProductInstanceUri/type
  .RevisionCounter should be >= -2147483648
    Received: <number> -2147483649
    Path:     mam.schema.jsonint32.schema.json/minimum
  .SerialNumber should be string
    Received: <null>
    Path:     mam.schema.json#/properties/SerialNumber/type
  .SoftwareRevision should be string
    Received: <null>
    Path:     mam.schema.json#/properties/SoftwareRevision/type
"
`;

exports[`mam schema (2) match fails for invalid config -> invalid property Foobar: invalid property Foobar 1`] = `
"expect(received).toMatchSchema(schema)

received
  should NOT have additional properties, but found 'Foobar'
    Expected: <boolean> false
    Path:     mam.schema.json#/additionalProperties
  .RevisionCounter should be <= 2147483647
    Received: <number> 2147483648
    Path:     mam.schema.jsonint32.schema.json/maximum
"
`;

exports[`mam schema (3) match fails for invalid config -> revision counter exceeds int32: revision counter exceeds int32 1`] = `
"expect(received).toMatchSchema(schema)

received
  .RevisionCounter should be <= 2147483647
    Received: <number> 2147483648
    Path:     mam.schema.jsonint32.schema.json/maximum
"
`;
